keps/sig-cloud-provider/2025-extend-konnectivity-for-both-directions/README.md

```diff
@@ -0,0 +1,678 @@
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+The goal of this proposal is to allow traffic to flow from the Node Network to the Master Network.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+API server network proxy has been originally introduced to allow running the cluster nodes on distinct isolated networks with respect to the one hosting the control plane components. This provides a way to handle traffic originating from the Kube API Server and going to the node networks. When using this setup, there are no other options than to directly expose the KAS to the Internet or setting up a VPN to handle traffic originated from the cluster nodes (i.e. Kubelet, pods).
+Exposing Konnectivity Proxy Server only would add an additional layer of security, given that the tunnels with proxy agents are secured with mTLS or Token authentication. This would protect it, for instance, from KAS misconfigurations or vulnerabilities that could expose sensitive information and/or access to unauthenticated users.
+Furthermore there is no standard solution for this kind of setup. It is possible to rely on VPNs for example to achieve a similar goal, but this requires specific implementations. What we propose here is to build on top of what we have, and having a consistent approach for master to node and node to master communications.
+Extending the API server network proxy to allow handling traffic in both directions seems to be the most consistent approach to address this issue. 
+Moreover, this will provide the possibility of routing agents’ traffic to their dedicated KAS based on SNI, enabling the option of load balancing traffic directed to different clusters with a single load balancer.
```

Comment:
Not sure where the best place to add the restrictions but I think we should detail some of the restrictions with this approach. Short list of the restrictions I am aware of.
1) APIServer Network Proxy Agent (Konnectivity Agent) cannot be deployed using DaemonSet in this model as Kubelet would be relying on the agent to get the set of pods to run. So the agent would need to be run as something like a static pod or a system service.
2) Nothing on the node is going to work properly until the agent comes up for the first time. Will need to confirm that retry logic in things like the Kubelet and Network Proxy (not to be confused with the APIServer Network Proxy) requests to the APIServer properly handle retry until the agent is working.

(Note I believe the current auth flows are just fine as either the SA token or the client cert should already be deployed with the agent.)

Reply:
Ok, makes sense we will integrate your points into the proposal.

Reply:
@cheftako assuming the agent is delivered as a static pod or a system service, I think it should be fine in the kubelet.  I have similar experience using a client side HAProxy to do similar that is managed by a static pod.